log-manager/src/main/java/io/airlift/log/RollingFileHandler.java

```diff
@@ -0,0 +1,579 @@
+                currentOutputStream.close();
+            }
+            catch (IOException e) {
+                reportError("Could not close output stream", e, CLOSE_FAILURE);
+            }
+            currentOutputFile = null;
+            currentOutputFileName = null;
+            currentOutputStream = null;
+            currentFileSize = 0;
+        }
+
+        // there shouldn't be any queued methods, but be safe
+        queue.clear();
+    }
+
+    private void processQueue()
+    {
+        List<byte[]> batch = new ArrayList<>(MAX_BATCH_COUNT);
+        boolean poisonMessageSeen = false;
+        while (!closed.get() || !poisonMessageSeen) {
+            if (queue.isEmpty()) {
+                try {
+                    batch.add(queue.take());
+                }
+                catch (InterruptedException ignored) {
+                }
+            }
+            else {
+                queue.drainTo(batch, MAX_BATCH_COUNT);
+            }
+
+            int poisonMessageIndex = getPoisonMessageIndex(batch);
+            if (poisonMessageIndex >= 0) {
+                poisonMessageSeen = true;
+                batch = batch.subList(0, poisonMessageIndex);
+            }
+
+            logMessageBatch(batch);
+            batch.clear();
+        }
+    }
+
+    private int getPoisonMessageIndex(List<byte[]> messages)
+    {
+        for (int i = 0; i < messages.size(); i++) {
+            if (messages.get(i) == poisonMessage) {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    private synchronized void logMessageBatch(List<byte[]> batch)
+    {
+        for (byte[] message : batch) {
+            if (currentFileSize + message.length > maxFileSize) {
+                try {
+                    rollFile();
+                }
+                catch (IOException e) {
+                    // it is possible the roll worked, but there was a problem cleaning up, and it is possible it failed
+                    // either way, roll will be attempted again once file grows by maxFileSize
+                    currentFileSize = 0;
+
+                    reportError("Error rolling log file", e, GENERIC_FAILURE);
+                }
+            }
+
+            logHistoryManager.pruneLogFilesIfNecessary(currentFileSize + message.length);
+
+            currentFileSize += message.length;
+            try {
+                currentOutputStream.write(message);
+            }
+            catch (Exception e) {
+                reportError(null, e, WRITE_FAILURE);
+            }
+        }
+        // always flush at the end of a batch, so logs aren't delayed
+        flush();
+    }
+
+    private synchronized void rollFile()
+            throws IOException
+    {
+        // carefully update the stream, so if there is a problem logging can continue
+
+        LogFileName newFileName = null;
+        Path newFile = null;
+        OutputStream newOutputStream = null;
+        for (int i = 0; i < 100; i++) {
```

Comment:
Is the purpose of this loop to retry if we see a `FileAlreadyExistsException`? In that case, think it will be ineffective. It will run all the iterations in an instant, so if the intent is to wait until the file "magically" disappears, there won't be enough time for anything meaningful to happen in between iterations. 

Also, is there any significance to the `100` limit? Maybe make it a constant for self-documentation purposes.

Reply:
Added a constant `MAX_OPEN_NEW_LOG_ATTEMPTS`